README.md
![logo](./build/github.jpg)

# UnPixel

Help your eyes take a break

![screenshot](./build/screenshot.png)

## 🤔 Why?

When staring for long times at computer screen our eyes can get dry and suffer form what is called Computer Vision Syndrome (CVS).

To combat and alleviate the symptoms there is the famous 20/20/20 rule that aims at reducing the stress on our eyes caused by focusing for long periods on the same distance.

This is not medical advice. Read more [here](https://en.wikipedia.org/wiki/Computer_vision_syndrome) and [here](https://www.aoa.org/healthy-eyes/eye-and-vision-conditions/computer-vision-syndrome).

## 💻 Installation

Head to the [release page](https://github.com/cupcakearmy/unpixel/releases) and grab the latest for your platform.

- For `macOS` you can downlaod the `.dmg` image.
- For `Windows` download the `.exe`.
- For `Linux` either the `.AppImage` or `.deb`, you will know what fits you 😉.

## 🛠 Building / Development

1. Clone the repo
2. `pnpm install`
3. `pnpm dist` to build for all platforms

Alternatively you can run `pnpm build` and the `pnpm electron-builder -m` for `macOS`, `-l` fir `linux` or `-w` for `windows`.

## 📦 Release

1. Bump version in `package.json`
2. Update `CHANGELOG.md`
3. Create a draft release with the new version. (e.g. tag with `v1.2.3`).
4. Start the publish action and the ci will build and upload assets to the draft matching the version of `package.json` and draft.


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [1.3.2] - 2022-01-21

### Changed

- Moved to pnpm

### Security

- Updated all dependencies.

## [1.3.1] - 2021-06-17

### Fixed

- Pause on Mic/Camera was not working correctly.

## [1.3.0] - 2021-06-10

### Added

- [macOS] Pause on Mic/Camera.

## [1.2.0] - 2021-06-01

### Added

- Progress bar.

## [1.1.2] - 2021-05-31

### Fixed

- Bug with update window when banner is currently open.

## [1.1.1] - 2021-05-28

### Added

- Restore focus to previous window when closing banner.
- Responsive image.

### Fixed

- Settings page not opening.

## [1.1.0] - 2021-05-15

### Added

- Only pause for a limited time.
- Volume slider for the sound.


src/back/tray.ts
import { Tray, Menu, nativeImage } from 'electron'
import path from 'path'
import ms from 'ms'

import Banner from './banner'
import Settings from './settings'
import { InputDevicesStatus } from './utils'

export default class TrayUtility {
  static tray: Tray | null = null

  static build() {
    const [paused, interval] = Settings.getStatus()
    let status = paused ? `Paused for: ${ms(interval)}` : `Next break: ${ms(interval)}`
    if (InputDevicesStatus.areCameraOrMicrophoneActive()) status = `Paused: Mic/Camera Active`

    const template: Parameters<typeof Menu['buildFromTemplate']>[0] = [
      { label: status, type: 'normal', enabled: false },
      { type: 'separator' },
      {
        label: 'Take a break now',
        type: 'normal',
        click: () => Banner.open(),
      },
    ]

    template.push(
      paused
        ? {
            label: 'Break pause',
            click: () => {
              Settings.save('paused', 0)
              this.build()
            },
          }
        : {
            label: 'Pause for...',
            submenu: Menu.buildFromTemplate(
              // Minutes to pause
              [10, 30, 60, 120, 360]
                .map((minutes) => minutes * 60 * 1000)
                .map((time) => ({
                  label: ms(time),
                  click: () => {
                    Settings.save('paused', Date.now() + time)
                    this.build()
                  },
                }))
            ),
          }
    )

    template.push(
      { label: 'Settings', click: () => Settings.open() },
      { type: 'separator' },
      { label: 'Quit', role: 'quit' }
    )

    const menu = Menu.buildFromTemplate(template)
    this.tray?.setContextMenu(menu)
  }

  static init() {
    if (!this.tray) {
      const file = path.join(__dirname, '../../assets/tray.png')
      const resized = nativeImage.createFromPath(file).resize({ width: 24, height: 24 })
      resized.setTemplateImage(true)
      resized.isMacTemplateImage = true
      this.tray = new Tray(resized)
      this.build()
    }
  }
}


src/back/index.ts
import { app } from 'electron'
import logger from 'electron-log'

import TrayUtility from './tray'
import Settings from './settings'
import Banner from './banner'
import Updater from './updater'
import { InputDevicesStatus } from './utils'

export const DEV = !app.isPackaged

// Enforce single instance
if (!app.requestSingleInstanceLock()) {
  logger.warn('Another instance is already running. Exiting.')
  app.quit()
}

// Disable gpu
app.disableHardwareAcceleration()
app.commandLine.appendSwitch('disable-software-rasterizer')

logger.catchErrors({ showDialog: true })
logger.log('Starting')
app
  .whenReady()
  .then(() => {
    logger.log('Initializing')
    if (!DEV) app.dock?.hide()
    TrayUtility.init()
    Settings.init()
    Banner.init()
    Updater.init()
    InputDevicesStatus.init()
    logger.log('Done')
  })
  .catch((e) => {
    logger.error(e)
    process.exit(1)
  })

app.on('window-all-closed', () => {
  // Prevent closing of the app
})


src/back/utils.ts
import cp from 'child_process'
import { promisify } from 'util'

import Settings from './settings'

const exec = promisify(cp.exec)

export async function isCameraActive(): Promise<boolean> {
  // Not reliable
  if (process.platform === 'darwin') {
    // Check number of processes using the camera
    const out = await exec(`lsof -n | grep "AppleCamera"`)
    const processesUsingCamera = out.stdout.trim().split('\n').length
    return processesUsingCamera > 1 // One is the apple daemon that is always active
  }
  return false
}

export async function isMicrophoneActive(): Promise<boolean> {
  if (process.platform === 'darwin') {
    const out = await exec(`ioreg -c AppleHDAEngineInput | grep "IOAudioEngineState"`)
    const parsed = parseInt(out.stdout.trim().replace(/[^\d]/gim, ''))
    return parsed > 0
  }
  return false
}

export class InputDevicesStatus {
  static status = {
    mic: false,
    camera: false,
  }

  static update() {
    // TODO: Update electron version as soon as issue is resolved https://github.com/electron/electron/issues/26143
    isMicrophoneActive().then((result) => (InputDevicesStatus.status.mic = result))
    isCameraActive().then((result) => (InputDevicesStatus.status.camera = result))
  }

  static init() {
    setInterval(InputDevicesStatus.update, 2000)
  }

  static areCameraOrMicrophoneActive(): boolean {
    if (Settings.load('skipOnCameraOrMic')) {
      return InputDevicesStatus.status.mic
    }
    return false
  }
}


src/back/banner.ts
import { join } from 'path'
import os from 'os'
import { app, BrowserWindow, BrowserWindowConstructorOptions, ipcMain } from 'electron'

import { DEV } from '.'
import Settings from './settings'
import TrayUtility from './tray'
import { InputDevicesStatus } from './utils'

export default class Banner {
  static interval: ReturnType<typeof setInterval>
  static window: BrowserWindow | null = null

  static init() {
    if (this.interval) return
    this.interval = setInterval(this.check, 1000)
    this.check()
    ipcMain.on('close', () => {
      this.close()
    })
  }

  static shouldShow(): boolean {
    const [paused, interval] = Settings.getStatus()
    if (paused || interval > 1000) return false
    if (InputDevicesStatus.areCameraOrMicrophoneActive()) return false
    return true
  }

  static check() {
    TrayUtility.build()
    if (Banner.shouldShow()) Banner.open()
  }

  static open() {
    if (this.window) return

    const options: BrowserWindowConstructorOptions = {
      frame: false,
      webPreferences: {
        nodeIntegration: true,
        contextIsolation: false,
      },
      width: 1200,
      height: 600,
      fullscreen: true,
      simpleFullscreen: os.platform() === 'darwin',
    }
    this.window = new BrowserWindow(options)

    const entry = join(__dirname, '../front/banner/index.html')
    this.window.loadFile(entry)

    if (DEV) {
      this.window.webContents.toggleDevTools()
    } else {
      this.window.setAlwaysOnTop(true, 'floating', 99)
      this.window.setVisibleOnAllWorkspaces(true, { visibleOnFullScreen: true })
      this.window.setMovable(false)
      this.window.setResizable(false)
      this.window.focus()
    }
  }

  static close() {
    if (this.window) {
      Settings.save('lastRun', Date.now())
      this.window.minimize()
      this.window.hide()
      if (process.platform === 'darwin') app.hide()
      this.window.close()
      this.window = null
    }
  }
}


src/back/settings.ts
import { BrowserWindow, ipcMain } from 'electron'
import Store from 'electron-store'
import { join } from 'path'
import AutoLaunch from 'auto-launch'

import { productName } from '../../package.json'

const autoLaunch = new AutoLaunch({ name: productName, mac: { useLaunchAgent: true } })

import { DEV } from '.'
import dayjs from 'dayjs'

const store = new Store()
const defaults = {
  every: 20,
  duration: 20,
  boot: true,
  paused: 0,
  lastRun: 0,
  autoClose: false,
  volume: 100,
  skipOnCameraOrMic: true,
}
export type SettingKeys = keyof typeof defaults
const IntNormalizer = (x: any) => parseInt(x)
const BoolNormalizer = (x: any) => !!x
const normalizers: Record<SettingKeys, (x: any) => any> = {
  every: IntNormalizer,
  duration: IntNormalizer,
  boot: BoolNormalizer,
  autoClose: BoolNormalizer,
  paused: IntNormalizer,
  lastRun: IntNormalizer,
  volume: IntNormalizer,
  skipOnCameraOrMic: BoolNormalizer,
}

export default class Settings {
  static win: BrowserWindow | null = null

  static init() {
    ipcMain.on('save', (e, { key, value }) => {
      this.save(key, value)
    })
    ipcMain.on('load', (e, { key }) => {
      e.returnValue = this.load(key)
    })
    if (Settings.load('boot')) {
      autoLaunch.enable()
    }
    Settings.save('lastRun', Date.now())
  }

  static save<T extends SettingKeys>(key: T, value: typeof defaults[T]) {
    const normalized = normalizers[key](value)
    store.set(key, normalized)
    if (key === 'boot') {
      normalized ? autoLaunch.enable() : autoLaunch.disable()
    }
  }

  static load<T extends SettingKeys>(key: T) {
    const saved = store.get(key) as typeof defaults[T] | undefined
    return saved ?? defaults[key]
  }

  static open() {
    if (this.win) return
    this.win = new BrowserWindow({
      width: 400,
      height: 690,
      center: true,
      resizable: false,
      webPreferences: {
        nodeIntegration: true,
        contextIsolation: false,
      },
    })
    this.win.on('closed', () => (this.win = null))

    const entry = join(__dirname, '../front/settings/index.html')
    this.win.loadFile(entry)
    this.win.setMenu(null)

    if (DEV) {
      this.win.setSize(800, 485)
      this.win.setResizable(true)
      this.win.webContents.openDevTools()
    }
  }

  static getStatus(): [boolean, number] {
    const paused: number = this.load('paused')
    const now = Date.now()
    if (paused > now) {
      return [true, paused - now]
    }

    const every = Settings.load('every')
    const lastRun = Settings.load('lastRun')
    const diff = every * 60 * 1000 - dayjs(now).diff(dayjs(lastRun), 'ms')
    return [false, diff]
  }
}


src/back/updater.ts
import semver from 'semver'
import axios from 'axios'
import logger from 'electron-log'
import { dialog, shell } from 'electron'

import pkg from '../../package.json'
import Banner from './banner'

const current = semver.coerce(pkg.version)
const url = 'https://api.github.com/repos/cupcakearmy/unpixel/tags'
const interval = 1000 * 60 * 15 // 10 Minutes

export default class Updater {
  static init() {
    this.check()
  }

  static async check() {
    const { data } = await axios({
      method: 'get',
      url,
    })
    // parse tags and reverse sort them to get the highest
    const tags = data.map((d: any) => semver.coerce(d.name)).sort(semver.rcompare)
    const latest = tags[0]
    if (!current) throw new Error('Could not determine current version')
    if (semver.lt(current, latest)) {
      logger.info('New version available')
      Banner.close()
      dialog
        // @ts-ignore
        .showMessageBox(null, {
          title: 'Update available',
          message: 'A newer version is available, please download.',
          buttons: ['Download', 'Later'],
          cancelId: 1,
        })
        .then(({ response }) => {
          if (response === 0) {
            shell.openExternal('https://github.com/cupcakearmy/unpixel/releases')
          }
        })
    } else {
      logger.info('Up to date')
      setTimeout(() => this.check(), interval)
    }
  }
}
